Remotion/Data/DomainObjects/Mapping/TypeDefinition.cs

```diff
@@ -0,0 +1,248 @@
+
+      if (IsReadOnly)
+        throw new NotSupportedException($"Type '{Type.GetFullNameSafe()}' is read-only.");
+
+      CheckPropertyDefinitions(propertyDefinitions);
+
+      _propertyDefinitions = propertyDefinitions;
+      _propertyDefinitions.SetReadOnly();
+    }
+
+    public RelationEndPointDefinitionCollection GetRelationEndPointDefinitions ()
+    {
+      return _cachedRelationEndPointDefinitions.Value;
+    }
+
+    public IRelationEndPointDefinition? GetRelationEndPointDefinition (string propertyName)
+    {
+      ArgumentUtility.CheckNotNullOrEmpty(nameof(propertyName), propertyName);
+
+      return _cachedRelationEndPointDefinitions.Value[propertyName];
+    }
+
+    public IRelationEndPointDefinition GetMandatoryRelationEndPointDefinition (string propertyName)
+    {
+      ArgumentUtility.CheckNotNullOrEmpty(nameof(propertyName), propertyName);
+
+      var relationEndPointDefinition = GetRelationEndPointDefinition(propertyName);
+      if (relationEndPointDefinition == null)
+        throw new MappingException($"No relation found for type '{Type.GetFullNameSafe()}' and property '{propertyName}'.");
+
+      return relationEndPointDefinition;
+    }
+
+    public void SetRelationEndPointDefinitions (RelationEndPointDefinitionCollection relationEndPoints)
+    {
+      ArgumentUtility.CheckNotNull(nameof(relationEndPoints), relationEndPoints);
+
+      if (_relationEndPoints != null)
+        throw new InvalidOperationException($"The relation end point definitions for type '{Type.GetFullNameSafe()}' have already been set.");
+
+      if (IsReadOnly)
+        throw new NotSupportedException($"Type '{Type.GetFullNameSafe()}' is read-only.");
+
+      CheckRelationEndPointDefinitions(relationEndPoints);
+
+      _relationEndPoints = relationEndPoints;
+      _relationEndPoints.SetReadOnly();
+    }
+
+    public PropertyDefinition? ResolveProperty (IPropertyInformation propertyInformation)
+    {
+      ArgumentUtility.CheckNotNull(nameof(propertyInformation), propertyInformation);
+
+      var propertyAccessorData = PropertyAccessorDataCache.ResolvePropertyAccessorData(propertyInformation);
+      return propertyAccessorData?.PropertyDefinition;
+    }
+
+    public IRelationEndPointDefinition? ResolveRelationEndPoint (IPropertyInformation propertyInformation)
+    {
+      ArgumentUtility.CheckNotNull(nameof(propertyInformation), propertyInformation);
+
+      var propertyAccessorData = PropertyAccessorDataCache.ResolvePropertyAccessorData(propertyInformation);
+      return propertyAccessorData?.RelationEndPointDefinition;
+    }
+
+    public bool IsAssignableFrom (TypeDefinition other)
+    {
+      ArgumentUtility.CheckNotNull(nameof(other), other);
+
+      return Type.IsAssignableFrom(other.Type);
+    }
+
+    public void SetReadOnly ()
+    {
+      IsReadOnly = true;
+    }
+
+    public override string ToString ()
+    {
+      return $"{GetType().GetFullNameChecked()}: {Type.GetFullNameSafe()}";
```

Comment:
Depending on what is done in ClassDefinition, we'd use the Name instead of FullName here, too. And do you want to do Expression Bodied members for the one liners?